scripts/licenses/save.go

```diff
@@ -0,0 +1,153 @@
+	// overwriteSavePath controls behaviour when the directory indicated by savepath already exists.
+	// If true, the directory will be replaced. If false, the command will fail.
+	overwriteSavePath bool
+
+	saveCmd = &cobra.Command{
+		Use:   "save <package>",
+		Short: "Saves licenses, copyright notices and source code, as required by a Go package's dependencies, to a directory.",
+		Args:  cobra.ExactArgs(1),
+		RunE:  saveMain,
+	}
+)
+
+func init() {
+	saveCmd.Flags().StringVar(&savePath, "save_path", "", "Directory into which files should be saved that are required by license terms")
+	if err := saveCmd.MarkFlagRequired("save_path"); err != nil {
+		glog.Fatal(err)
+	}
+	if err := saveCmd.MarkFlagFilename("save_path"); err != nil {
+		glog.Fatal(err)
+	}
+
+	saveCmd.Flags().BoolVar(&overwriteSavePath, "force", false, "Delete the destination directory if it already exists.")
+
+	rootCmd.AddCommand(saveCmd)
+}
+
+func saveMain(cmd *cobra.Command, args []string) error {
+	importPath := args[0]
+
+	if overwriteSavePath {
+		if err := os.RemoveAll(savePath); err != nil {
+			return err
+		}
+	}
+
+	// Check that the save path doesn't exist, otherwise it'd end up with a mix of
+	// existing files and the output of this command.
+	if d, err := os.Open(savePath); err == nil {
+		d.Close()
+		return fmt.Errorf("%s already exists", savePath)
+	} else if !os.IsNotExist(err) {
+		return err
+	}
+
+	classifier, err := licenses.NewClassifier(confidenceThreshold)
+	if err != nil {
+		return err
+	}
+
+	var pkgsWithUnknownLicense []*build.Package
+	libs, err := libraries(importPath)
+	if err != nil {
+		return err
+	}
+	for _, lib := range libs {
+		libSaveDir := filepath.Join(savePath, unvendor(lib.Name()))
+		// Detect what type of license this library has and fulfill its requirements, e.g. copy license, copyright notice, source code, etc.
+		licenseName, licenseType, err := classifier.Identify(lib.LicensePath)
+		if err != nil {
+			return err
+		}
+		switch licenseType {
+		case licenses.Restricted, licenses.Reciprocal:
+			// Copy the entire source directory for the library.
+			libDir := filepath.Dir(lib.LicensePath)
+			if err := copySrc(libDir, libSaveDir); err != nil {
+				return err
+			}
+		case licenses.Notice, licenses.Permissive, licenses.Unencumbered:
+			// Just copy the license and copyright notice.
+			if err := copyNotices(lib.LicensePath, libSaveDir); err != nil {
+				return err
+			}
+		case licenses.Forbidden:
+			return fmt.Errorf("forbidden license %q used by this library: %s", licenseName, lib)
+		case licenses.Unknown:
+			pkgsWithUnknownLicense = append(pkgsWithUnknownLicense, lib.Packages...)
+		default:
+			glog.Fatalf("%q license is of an unrecognised type: %q", licenseName, licenseType)
+		}
+	}
+	if len(pkgsWithUnknownLicense) > 0 {
+		return fmt.Errorf("Packages with unknown license: %v", pkgsWithUnknownLicense)
+	}
+	return nil
+}
+
+func copySrc(src, dest string) error {
+	if err := copy.Copy(src, dest); err != nil {
```

Comment:
Hmm, just wondered whether we should worry about un-checked-in files in the local Git repo getting swept up here.  (Not sure I can think of what to do about it, mind).

Reply:
We could check for that by doing the equivalent of `git status` on each repo and returning an error for a dirty repo. Alternatively, we could simply print a warning about this when running the save command.